Xilinx/finn
# Feature/linear collapse repeated

## Dependencies:
The new code in this PR doesn't strictly depend't on #113 
However, I needed reorder transforms to do some checks, reason why I this is based from the linear reorder transforms branch
Once those are merged, this PR will only consist of 2 commits

## Changes in code
Add checks to `CollapseRepeatedOp`. Check that:
- neither of the nodes to merge were joins 
- there was no fork between those nodes

## Tests
- Add check to previous test to verify that the transformation is doing the required job
- Add test to verify that CollapseRepeatedOp only operates on linear segments.
